### ISSUE TITLE
Change `prefer-const` to permit the author to use `let` in the case of declaring an object that is then mutated.

### ISSUE THREAD

**Issue by gausie** [eslint/eslint#6152]
**Version**
v2.9.0

**Problem you want to solve**
As a developer, I might want to choose to use `let` when declaring an object I later mutate, even though JavaScript will allow me to use `const`. This would be to make the status of the variable clearer to future readers and myself. At the moment, it is not possible to do this while using `prefer-const`, which is an important rule in other situations.

**My take on the solution**
The parser should look to see if a variable is mutated in the scope and, if so, permit the use of `let` for that variable. `prefer-const` should have an option that enables this behaviour.

**What you did**
With `prefer-const` enabled, write
```javascript
function example(value) {
  let object = {};
  object.property = value;
  return object;
}
```

**What you would like to happen**
No errors to be triggered

**What happened**
The second line is highlighted by `prefer-const`.

**Comment by Canop** [comment]
You didn't change the value of the `object` reference. Using `const` for the declaration wouldn't prevent changing a property of the (unchanged) value of `object`.

**Comment by mysticatea** [comment]
Thank you for this issue.

Hmm, personally I want not add this option.
`const` of JavaScript doesn't mean that object's properties are not modified.
If people rely on `const` keyword to check whether properties are not modified or not, it's wrong. So this option would mislead people, I think.

**Comment by gausie** [comment]
I understand that JavaScript will not stop you from modifying an object's properties if you use `const`. This is more to support a developer/team's decision that they might want to indicate (for code readability) that they intend for an object's properties to be changed by using `let` instead of `const`.

I definitely don't think this should be the default behaviour, but it seems like a reasonable case to be controlled by an option.

**Comment by CodyReichert** [comment]
I agree with @mysticatea that not throwing an error would be misleading - since that's how JavaScript works. I can see the benefit of having it as an option, though. Maybe an option on `prefer-const`, say `property-assignment`, that could be set to `always` or `never` would be helpful.

As far as the default behavior of the rule - it should match the JS specification.

**Comment by ilyavolodin** [comment]
This is not something we can do with static analysis 100% correctly. Here's an example:

```js
let temp = require('somelibrary');

let a = {};
temp(a);
```

No way for us to track that `temp` will modify a property of an object.

**Comment by alberto** [comment]
Thanks for your interest in improving eslint. Unfortunately, it looks like consensus couldn't be reached on this issue and so I'm closing it. While we wish we'd be able to accommodate everyone's requests, we do need to prioritize. We've found that issues failing to reach consensus after 21 days tend never to reach consensus, and as such, we close those issues. This doesn't mean the idea isn't interesting, just that it's not something the team can commit to.

**Comment by unleashit** [comment]
I agree with the OP. In my opinion, a constant is something that remains... constant. I don't personally agree with the current spec which allows object properties of constants to be mutable. Many before me have also felt there are "good parts" and bad parts of Javascript and have decided to choose a slightly different workflow than the authors intended.

Just my 2cents. I'm not able to use this rule because in my workflow it gives me false positives, but I appreciate the great work of eslint!

**Comment by platinumazure** [comment]
Note that if someone wants to implement a "don't mutate const-declared objects" rule, s/he is welcome to do so in a plugin! :smile:

**Comment by unleashit** [comment]
@platinumazure I'm one of the someones ;-)

Sadly it's a bit over my head especially considering the amount of time I currently have. Just wanted to give this my plus 1 and state that I see good code by good developers "all the time" who treat const as immutable, despite the the annoyance of the spec.
